### 下載Youtube影片
ytDloadMergeV0.3.ipynb --> ~2024.06.20止
ytDloadMergeV0.4.ipynb -->  2024.06.20~ 起
Pytube：Youtube影片下載

In [1]:
# 使用模組
from pytube import YouTube

url='https://youtube.com/live/o6HQepxxDlo'   # 2023_11_10_下午
yt = YouTube(url)

# 影片格式分析
print(f'標題: {yt.title}')              # 影片標題
print(f'長度(s): {yt.length}')          # 影片長度 ( 秒 )
print(f'作者: {yt.author}')             # 影片作者
print(f'網址: {yt.channel_url}')        # 影片作者頻道網址
print(f'縮圖網址: {yt.thumbnail_url}')  # 影片縮圖網址
print(f'觀看數: {yt.views}')            # 影片觀看數

# 分析
print(f'支援畫質:{yt.streams}')  # 影片支援哪些畫質
# print(yt.streams.all())  # all()-->useless
print('影片支援畫質數: ',len(yt.streams))

for st in yt.streams:
  print(st)


標題: 2023_11_10_下午
長度(s): 10055
作者: 徐國堂
網址: https://www.youtube.com/channel/UCKd_tuM8M79b1GAwAlTi9ZA
縮圖網址: https://i.ytimg.com/vi/o6HQepxxDlo/hq720.jpg?sqp=-oaymwEmCIAKENAF8quKqQMa8AEB-AH-CYAC0AWKAgwIABABGGUgZShlMA8=&rs=AOn4CLDfukQ0xJCh4CLKvLbNK4aegmeIQg
觀看數: 30
支援畫質:[<Stream: itag="18" mime_type="video/mp4" res="360p" fps="30fps" vcodec="avc1.42001E" acodec="mp4a.40.2" progressive="True" type="video">, <Stream: itag="137" mime_type="video/mp4" res="1080p" fps="30fps" vcodec="avc1.640028" progressive="False" type="video">, <Stream: itag="136" mime_type="video/mp4" res="720p" fps="30fps" vcodec="avc1.64001f" progressive="False" type="video">, <Stream: itag="134" mime_type="video/mp4" res="360p" fps="30fps" vcodec="avc1.4d401e" progressive="False" type="video">, <Stream: itag="160" mime_type="video/mp4" res="144p" fps="30fps" vcodec="avc1.4d400c" progressive="False" type="video">, <Stream: itag="139" mime_type="audio/mp4" abr="48kbps" acodec="mp4a.40.5" progressive="False" type="audio">, 

一下為<參考> of ytDownLoadMergeAv

In [2]:
# 用itag分別下載聲音&影像  _ ok
## audio(itag=140) 音訊取樣率44KHz
## vedio(itag=137) 為1080p
## 使用 yt.title 為檔名
from pytube import YouTube

url='https://youtube.com/live/o6HQepxxDlo'   # 2023_11_10_下午
yt = YouTube(url) 

yt_title=yt.title
itag_Audio = 140
itag_Vedio = 137
itag_List = [itag_Audio,itag_Vedio]
for i in itag_List:
    stream = yt.streams.get_by_itag(i)
    stream.download(output_path='C:/Users/Lu/Documents/youtube',filename=f'python_{yt_title}_itag({i}).mp4', skip_existing=True, max_retries=0)
    print(f'python_{yt_title}_itag({i})完成')
print('downdload完成')

python_2023_11_20_下午_itag(140)完成
python_2023_11_20_下午_itag(137)完成
downdload完成


In [8]:
import os

os.remove(f'C:/Users/Lu/Documents/youtube/python_{yt_title}_itag({itag_Vedio}).mp4')
os.remove(f'C:/Users/Lu/Documents/youtube/python_{yt_title}_itag({itag_Audio}).mp4')
print('影像&音訊檔案刪除成功,合併完成！')

影像&音訊檔案刪除成功,合併完成！


In [2]:
# V0.4.0 -->完成 (2024.06.15止-->final)
from pytube import YouTube
import subprocess
import os

def media_contents(url):
    # 影片格式分析
    print('標題: ',yt.title)                 # 影片標題
    print('長度 ( 秒 ): ',yt.length)         # 影片長度 ( 秒 )
    print('作者: ',yt.author)                # 影片作者
    print('作者頻道網址: ',yt.channel_url)   # 影片作者頻道網址
    print('縮圖網址: ',yt.thumbnail_url)     # 影片縮圖網址
    print('觀看數: ',yt.views)               # 影片觀看數

    # 分析
    #print(yt.streams)  # 影片支援哪些畫質
    print('stream數: ',len(yt.streams))

    for st in yt.streams:
        print(st)
    print('media格式分析完成')


def download_video_and_audio(url,itag_Vedio,itag_Audio):
    print('downdloading of video & audio .....')
    itag_List = [itag_Vedio,itag_Audio]
    for i in itag_List:
        stream = yt.streams.get_by_itag(i)
        stream.download(output_path='C:/Users/Lu/Documents/youtube',filename=f'python_{yt.title}_itag({i}).mp4', skip_existing=True, max_retries=0)
        print(f'python_{yt.title}_itag({i})完成')
    print('downdload完成')


def merge_video_and_audio(video_path, audio_path, output_path):
    command = [
        'ffmpeg',
        '-i', video_path,
        '-i', audio_path,
        '-c', 'copy',
        '-map', '0:v',
        '-map', '1:a',
        output_path
    ]
    
    try:
        print("執行merge命令:", ' '.join(command))  # 增加執行內容訊息
        result = subprocess.run(command, check=True, stdout=subprocess.PIPE, stderr=subprocess.PIPE, text=True, encoding='utf-8',timeout=30)
        print("合併成功！輸出文件位於:", output_path)
    except subprocess.CalledProcessError as e:
        print("ffmpeg合併過程中發生錯誤:", e)
        if e.stderr:
            try:
                # 嘗試用 UTF-8 解碼
                print("ffmpeg命令輸出:", e.stderr)
            except UnicodeDecodeError:
                # 如果失敗，使用 CP950 解碼
                print("ffmpeg命令輸出:", e.stderr.decode(encoding='cp950', errors='ignore'))
        else:
            print("無法獲取 ffmpeg 錯誤輸出。")
    except subprocess.TimeoutExpired as e:
        print("命令執行超時:", e)


#url = input("Enter the url of media : ")   
url='https://youtube.com/live/o6HQepxxDlo'   # 2023_11_10_下午
yt = YouTube(url) 
itag_Audio = 140
itag_Vedio = 137 
media_contents(url)
download_video_and_audio(url,itag_Vedio,itag_Audio)
merge_video_and_audio(f'C:/Users/Lu/Documents/youtube/python_{yt.title}_itag({itag_Vedio}).mp4', f'C:/Users/Lu/Documents/youtube/python_{yt.title}_itag({itag_Audio}).mp4', f'C:/Users/Lu/Documents/youtube/python_{yt.title}_itag(merge).mp4')

os.remove(f'C:/Users/Lu/Documents/youtube/python_{yt.title}_itag({itag_Vedio}).mp4')
os.remove(f'C:/Users/Lu/Documents/youtube/python_{yt.title}_itag({itag_Audio}).mp4')
print('合併成功！影像&音訊檔案刪除完成')

#<2023_11_10_下午> media 長度 03:10:30 容量大小731,905KB 處理耗時10m 6.4s

標題:  2023_11_10_下午
長度 ( 秒 ):  10055
作者:  徐國堂
作者頻道網址:  https://www.youtube.com/channel/UCKd_tuM8M79b1GAwAlTi9ZA
縮圖網址:  https://i.ytimg.com/vi/o6HQepxxDlo/hq720.jpg?sqp=-oaymwEmCIAKENAF8quKqQMa8AEB-AH-CYAC0AWKAgwIABABGGUgZShlMA8=&rs=AOn4CLDfukQ0xJCh4CLKvLbNK4aegmeIQg
觀看數:  30
stream數:  8
<Stream: itag="18" mime_type="video/mp4" res="360p" fps="30fps" vcodec="avc1.42001E" acodec="mp4a.40.2" progressive="True" type="video">
<Stream: itag="137" mime_type="video/mp4" res="1080p" fps="30fps" vcodec="avc1.640028" progressive="False" type="video">
<Stream: itag="136" mime_type="video/mp4" res="720p" fps="30fps" vcodec="avc1.64001f" progressive="False" type="video">
<Stream: itag="134" mime_type="video/mp4" res="360p" fps="30fps" vcodec="avc1.4d401e" progressive="False" type="video">
<Stream: itag="160" mime_type="video/mp4" res="144p" fps="30fps" vcodec="avc1.4d400c" progressive="False" type="video">
<Stream: itag="139" mime_type="audio/mp4" abr="48kbps" acodec="mp4a.40.5" progressive="False" t

========================================================================================
##### 以下為get_audio_only相關


In [11]:
# v0.1 for test

from pytube import YouTube

yt = YouTube('https://www.youtube.com/watch?v=aGZk0nEBWv4')   # Soft Classical Piano - Relaxing Piano Pieces (Luke Faulkner)
# yt = YouTube(url) 

yt_audio = yt.streams.filter(abr='128kbps').first()
yt_audio.download()

print('downdload完成')

downdload完成


In [1]:
# V0.2 for test

from pytube import YouTube

yt = YouTube('https://www.youtube.com/watch?v=U_oxOhncGe0&t=9456s')   # 最好的古典音樂2023🎼古典音樂，學習和放鬆。莫扎特，貝多芬
# yt = YouTube(url) 

yt_audio = yt.streams.filter(abr='128kbps').first()
# yt_audio.download()

yt_audio.download(output_path='C:/Users/Lu/Documents/youtube',filename=f'yt_audio_{yt.title}.mp4', skip_existing=True, max_retries=0)

print('downdload完成')



downdload完成


In [2]:
# V0.3 for test

from pytube import YouTube

yt = YouTube('https://www.youtube.com/watch?v=IgeF8JfotNw')   # 世界名曲（流浪之歌-聖母頌-天鵝-美的羅斯瑪玲-幽默曲-小步舞曲-夢幻曲-愛之喜-布拉姆斯小夜曲-歎息小夜曲-古諾小夜曲-海頓小夜曲-杜立.......)
# yt = YouTube(url) 

yt_audio = yt.streams.filter(abr='128kbps').first()
# yt_audio.download()

yt_audio.download(output_path='C:/Users/Lu/Documents/youtube',filename=f'yt_audio_{yt.title}.mp4', skip_existing=True, max_retries=0)

print(f'downdload < {yt.title} > 完成')

downdload < 世界名曲（流浪之歌-聖母頌-天鵝-美的羅斯瑪玲-幽默曲-小步舞曲-夢幻曲-愛之喜-布拉姆斯小夜曲-歎息小夜曲-古諾小夜曲-海頓小夜曲-杜立歌小夜曲-小犬小夜曲-杜鵑小夜曲-皇帝圓舞曲-亦然音樂071） > 完成


In [2]:
# V0.4 for test

from pytube import YouTube

#url = input("Enter the url of media : ")   
url = 'https://www.youtube.com/watch?v=StfwZUc2DaQ'  # 曲目:Tchaikovsky最好的|有史以來最著名的古典音樂作品
yt = YouTube(url)   

yt_audio = yt.streams.filter(abr='128kbps').first()
# yt_audio.download()

yt_audio.download(output_path='C:/Users/Lu/Documents/youtube',filename=f'yt_audio_{yt.title}.mp4 \w', skip_existing=True, max_retries=0)
# yt_audio.download(output_path='C:/Users/Lu/Documents/youtube',filename=f'yt_audio.mp4', skip_existing=True, max_retries=0)  #執行ok

print(f'downdload < {yt.title} > 完成')

OSError: [Errno 22] Invalid argument: 'C:/Users/Lu/Documents/youtube\\yt_audio_The Best of Tchaikovsky | The Most Famous Classical Music Pieces of All Time.mp4 \\w'

In [14]:
# V0.41 for test

from pytube import YouTube

#url = input("Enter the url of media : ")   
url = 'https://www.youtube.com/watch?v=StfwZUc2DaQ'  # 曲目:Tchaikovsky最好的|有史以來最著名的古典音樂作品
yt = YouTube(url)   

yt_audio = yt.streams.filter(abr='128kbps').first()
# yt_audio.download()

# yt_audio.download(output_path='C:/Users/Lu/Documents/youtube',filename=f'yt_audio_{yt.title}.mp4', skip_existing=True, max_retries=0)   #執行NG-OSError: [Errno 22] Invalid argument:
# yt_audio.download(output_path='C:/Users/Lu/Documents/youtube',filename=f'yt_audio.mp4', skip_existing=True, max_retries=0)  #執行ok

# yt_title=yt.title[:15]
# yt_list=yt_title[:10]

#yt_audio.download(output_path='C:/Users/Lu/Documents/youtube',filename=f'yt_audio_{yt.title[:25]}.mp4', skip_existing=True, max_retries=0) #執行NG


yt_audio.download(output_path='C:/Users/Lu/Documents/youtube',filename=f'yt_audio_{yt.title[:23]}.mp4', skip_existing=True, max_retries=0)   #執行ok

print(f'downdload < {yt.title} > 完成')

downdload < The Best of Tchaikovsky | The Most Famous Classical Music Pieces of All Time > 完成


In [32]:
# V0.42 for test_1

from pytube import YouTube
import re

#url = input("Enter the url of media : ")   
url = 'https://www.youtube.com/watch?v=StfwZUc2DaQ'  # 曲目:Tchaikovsky最好的|有史以來最著名的古典音樂作品
yt = YouTube(url)   

#yt_titles = yt.title
print('全名 : ',yt.title)
yt_title = re.search(r'[^|]+',yt.title).group()
print('簡稱 : ',yt_title)


yt_audio = yt.streams.filter(abr='128kbps').first()

yt_audio.download(output_path='C:/Users/Lu/Documents/youtube',filename=f'yt_audio_{yt_title}.mp4', skip_existing=True, max_retries=0)   #執行ok

# yt_audio.download(output_path='C:/Users/Lu/Documents/youtube',filename=f'yt_audio_{yt.title[:23]}.mp4', skip_existing=True, max_retries=0)   #執行ok

print(f'downdload < {yt.title} > 完成')
# 需時 : 47.1s

全名 :  The Best of Tchaikovsky | The Most Famous Classical Music Pieces of All Time
簡稱 :  The Best of Tchaikovsky 
downdload < The Best of Tchaikovsky | The Most Famous Classical Music Pieces of All Time > 完成


In [36]:
# V0.42 for test_1.1

from pytube import YouTube
import re

#url = input("Enter the url of media : ")   
url = 'https://www.youtube.com/watch?v=StfwZUc2DaQ'  # 曲目:Tchaikovsky最好的|有史以來最著名的古典音樂作品
yt = YouTube(url)   

#yt_titles = yt.title
print('全名 : ',yt.title)
yt_title = re.search(r'[^|]*',yt.title).group()
print('簡稱 : ',yt_title)


yt_audio = yt.streams.filter(abr='128kbps').first()

yt_audio.download(output_path='C:/Users/Lu/Documents/youtube',filename=f'yt_audio_{yt_title}.mp4', skip_existing=True, max_retries=0)   #執行ok


print(f'downdload < {yt.title} > 完成')

全名 :  The Best of Tchaikovsky | The Most Famous Classical Music Pieces of All Time
簡稱 :  The Best of Tchaikovsky 
downdload < The Best of Tchaikovsky | The Most Famous Classical Music Pieces of All Time > 完成


In [34]:
# V0.42 for test_2

from pytube import YouTube
import re

#url = input("Enter the url of media : ")   
url = 'https://www.youtube.com/watch?v=VBNUwMd_eF4&t=4029s'  # 曲目:T蕭邦 一定要聽的鋼琴古典樂 ｜放鬆音樂 讀書 專注 The Best "Chopin" Classical Piano Music｜Relaxing Studying Music
yt = YouTube(url)   

#yt_titles = yt.title
print('全名 : ',yt.title)
yt_title = re.search(r'[^"|]+',yt.title).group()
print('簡稱 : ',yt_title)


yt_audio = yt.streams.filter(abr='128kbps').first()

yt_audio.download(output_path='C:/Users/Lu/Documents/youtube',filename=f'yt_audio_{yt_title}.mp4', skip_existing=True, max_retries=0)   #執行ok

# yt_audio.download(output_path='C:/Users/Lu/Documents/youtube',filename=f'yt_audio_{yt.title[:23]}.mp4', skip_existing=True, max_retries=0)   #執行ok

print(f'downdload < {yt.title} > 完成')

全名 :  The Best "Chopin" Classical Piano Music｜Relaxing Studying Music
簡稱 :  The Best 
downdload < The Best "Chopin" Classical Piano Music｜Relaxing Studying Music > 完成


In [43]:
# V0.42 for test_2.1

from pytube import YouTube
import re

#url = input("Enter the url of media : ")   
url = 'https://www.youtube.com/watch?v=VBNUwMd_eF4&t=4029s'  # 曲目:T蕭邦 一定要聽的鋼琴古典樂 ｜放鬆音樂 讀書 專注 The Best "Chopin" Classical Piano Music｜Relaxing Studying Music
yt = YouTube(url)   

print('全名 : ',yt.title)
yt_title = re.search(r'[^"|]+',yt.title).group()
print('簡稱 : ',yt_title)


yt_audio = yt.streams.filter(abr='128kbps').first()

yt_audio.download(output_path='C:/Users/Lu/Documents/youtube',filename=f'yt_audio_{yt_title}.mp4', skip_existing=True, max_retries=0)   #執行ok

print(f'downdload < {yt.title} > 完成')

全名 :  The Best "Chopin" Classical Piano Music｜Relaxing Studying Music
簡稱 :  The Best 
downdload < The Best "Chopin" Classical Piano Music｜Relaxing Studying Music > 完成


In [52]:
# V0.42 for test_3 for window file formate 

from pytube import YouTube
import re

#url = input("Enter the url of media : ")   
url = 'https://www.youtube.com/watch?v=VBNUwMd_eF4&t=4029s'  # 曲目:T蕭邦 一定要聽的鋼琴古典樂 ｜放鬆音樂 讀書 專注 The Best "Chopin" Classical Piano Music｜Relaxing Studying Music
yt = YouTube(url)   

#yt_titles = yt.title
print('全名 : ',yt.title)
#yt_title = re.search(r'[^"|]+',yt.title).group()
#yt_title = re.sub(r'[^|]+ |[^|]',r'_',yt.title)
# yt_title = re.search(r'[A-Za-z0-9 "]+',yt.title).group()  #OSError: [Errno 22] Invalid argument
yt_titles = re.sub('"','_',yt.title)
print(yt_titles)
yt_title = re.search(r'[A-Za-z0-9_ ]+',yt_titles).group()
# print(yt_titles)
print('簡稱 : ',yt_title)

yt_audio = yt.streams.filter(abr='128kbps').first()

yt_audio.download(output_path='C:/Users/Lu/Documents/youtube',filename=f'yt_audio_{yt_title}.mp4', skip_existing=True, max_retries=0)   #執行ok

print(f'downdload < {yt.title} > 完成')

全名 :  The Best "Chopin" Classical Piano Music｜Relaxing Studying Music
The Best _Chopin_ Classical Piano Music｜Relaxing Studying Music
簡稱 :  The Best _Chopin_ Classical Piano Music
downdload < The Best "Chopin" Classical Piano Music｜Relaxing Studying Music > 完成


In [67]:
# V0.42 for test_3.1 for window file formate 

from pytube import YouTube
import re

#url = input("Enter the url of media : ")   
url = 'https://www.youtube.com/watch?v=VBNUwMd_eF4&t=4029s'  # 曲目:T蕭邦 一定要聽的鋼琴古典樂 ｜放鬆音樂 讀書 專注 The Best "Chopin" Classical Piano Music｜Relaxing Studying Music
yt = YouTube(url)   

#yt_titles = yt.title
print('全名 : ',yt.title)
list=['/',':','"','<','>','\?','\*']
yt_titles = yt.title
for i in list:
    yt_title = re.sub(f'{i}','_',yt_titles)
    yt_titles = yt_title
    print(i)
    print(yt_titles)
#yt_title = re.search(r'[A-Za-z0-9_ ]+',yt_titles).group()
# print(yt_titles)
#print('簡稱 : ',yt_title)

#yt_audio = yt.streams.filter(abr='128kbps').first()

#yt_audio.download(output_path='C:/Users/Lu/Documents/youtube',filename=f'yt_audio_{yt_title}.mp4', skip_existing=True, max_retries=0)   #執行ok

#print(f'downdload < {yt.title} > 完成')

全名 :  The Best "Chopin" Classical Piano Music｜Relaxing Studying Music
/
The Best "Chopin" Classical Piano Music｜Relaxing Studying Music
:
The Best "Chopin" Classical Piano Music｜Relaxing Studying Music
"
The Best _Chopin_ Classical Piano Music｜Relaxing Studying Music
<
The Best _Chopin_ Classical Piano Music｜Relaxing Studying Music
>
The Best _Chopin_ Classical Piano Music｜Relaxing Studying Music
\?
The Best _Chopin_ Classical Piano Music｜Relaxing Studying Music
\*
The Best _Chopin_ Classical Piano Music｜Relaxing Studying Music


In [6]:
# V0.42 for test_3.2 for window file formate 

import re

ytTitle= '曲目:T蕭邦 一定要聽的鋼琴古典樂 ｜放鬆音樂 讀書 專注 The Best "Chopin" Classical Piano Music｜Relaxing Studying Music'

#print('全名 : ',ytTitle)
list=['/',':','"','<','>','\?','\*',' | ']
yt_titles = ytTitle
for i in list:
    yt_title = re.sub(f'{i}','_',yt_titles)
    yt_titles = yt_title
    print(i)
    print(yt_titles)

/
曲目:T蕭邦 一定要聽的鋼琴古典樂 ｜放鬆音樂 讀書 專注 The Best "Chopin" Classical Piano Music｜Relaxing Studying Music
:
曲目_T蕭邦 一定要聽的鋼琴古典樂 ｜放鬆音樂 讀書 專注 The Best "Chopin" Classical Piano Music｜Relaxing Studying Music
"
曲目_T蕭邦 一定要聽的鋼琴古典樂 ｜放鬆音樂 讀書 專注 The Best _Chopin_ Classical Piano Music｜Relaxing Studying Music
<
曲目_T蕭邦 一定要聽的鋼琴古典樂 ｜放鬆音樂 讀書 專注 The Best _Chopin_ Classical Piano Music｜Relaxing Studying Music
>
曲目_T蕭邦 一定要聽的鋼琴古典樂 ｜放鬆音樂 讀書 專注 The Best _Chopin_ Classical Piano Music｜Relaxing Studying Music
\?
曲目_T蕭邦 一定要聽的鋼琴古典樂 ｜放鬆音樂 讀書 專注 The Best _Chopin_ Classical Piano Music｜Relaxing Studying Music
\*
曲目_T蕭邦 一定要聽的鋼琴古典樂 ｜放鬆音樂 讀書 專注 The Best _Chopin_ Classical Piano Music｜Relaxing Studying Music
 | 
曲目_T蕭邦_一定要聽的鋼琴古典樂_｜放鬆音樂_讀書_專注_The_Best__Chopin__Classical_Piano_Music｜Relaxing_Studying_Music


In [19]:
# V0.42 for test_3.3 for window file formate 

import re
ytTitle= '曲目:T蕭邦/ 一定要聽的鋼琴古典樂? ｜<放鬆音樂> *讀書 專注 The Best "Chopin" Classical Piano Music｜Relaxing Studying Music'
print(type(ytTitle))

#print('全名 : ',ytTitle)
list=['/',':','"','<','>','\?','\*',' \|']
yt_titles = ytTitle
for i in list:
    yt_title = re.sub(f'{i}','_',yt_titles)
    yt_titles = yt_title
    print(i)
    print(yt_titles)

<class 'str'>
/
曲目:T蕭邦_ 一定要聽的鋼琴古典樂? ｜<放鬆音樂> *讀書 專注 The Best "Chopin" Classical Piano Music｜Relaxing Studying Music
:
曲目_T蕭邦_ 一定要聽的鋼琴古典樂? ｜<放鬆音樂> *讀書 專注 The Best "Chopin" Classical Piano Music｜Relaxing Studying Music
"
曲目_T蕭邦_ 一定要聽的鋼琴古典樂? ｜<放鬆音樂> *讀書 專注 The Best _Chopin_ Classical Piano Music｜Relaxing Studying Music
<
曲目_T蕭邦_ 一定要聽的鋼琴古典樂? ｜_放鬆音樂> *讀書 專注 The Best _Chopin_ Classical Piano Music｜Relaxing Studying Music
>
曲目_T蕭邦_ 一定要聽的鋼琴古典樂? ｜_放鬆音樂_ *讀書 專注 The Best _Chopin_ Classical Piano Music｜Relaxing Studying Music
\?
曲目_T蕭邦_ 一定要聽的鋼琴古典樂_ ｜_放鬆音樂_ *讀書 專注 The Best _Chopin_ Classical Piano Music｜Relaxing Studying Music
\*
曲目_T蕭邦_ 一定要聽的鋼琴古典樂_ ｜_放鬆音樂_ _讀書 專注 The Best _Chopin_ Classical Piano Music｜Relaxing Studying Music
 \|
曲目_T蕭邦_ 一定要聽的鋼琴古典樂_ ｜_放鬆音樂_ _讀書 專注 The Best _Chopin_ Classical Piano Music｜Relaxing Studying Music


In [20]:
# V0.42 for test_3.4 for window file formate --> pass
# How to escape special characters of a string with single backslashes [duplicate]
# https://stackoverflow.com/questions/18935754/how-to-escape-special-characters-of-a-string-with-single-backslashes

import re
ytTitle= '曲目:T蕭邦/ 一定要聽的鋼琴古典樂? ｜<放鬆音樂> *讀書 專注 The Best "Chopin" Classical Piano Music｜Relaxing Studying Music'
print(type(ytTitle))
list=['/',':','"','<','>','\?','\*',' \|']
escaped = ytTitle.translate(str.maketrans({"-":  r"\-",
                                           "]":  r"\]",
                                          "\\": r"\\",
                                           "^":  r"\^",
                                           "$":  r"\$",
                                           "*":  r"\*",
                                           ".":  r"\.",
                                           "|":  r"\|"}))
print(escaped)

<class 'str'>
曲目:T蕭邦/ 一定要聽的鋼琴古典樂? ｜<放鬆音樂> \*讀書 專注 The Best "Chopin" Classical Piano Music｜Relaxing Studying Music


=====================================================================================
+++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++
pass

In [29]:
# V0.5  --> pass
from pytube import YouTube
import re

#url = input("Enter the url of media : ")   
url = 'https://www.youtube.com/watch?v=StfwZUc2DaQ'  # 曲目:Tchaikovsky最好的|有史以來最著名的古典音樂作品
yt = YouTube(url)   

yt_title = yt.title
print(yt_title)

yt_titles = re.search(r'\D+',yt_title).group()
print(yt_titles)

yt_titles = re.search(r'[A-Za-z0-9 ]+',yt_title).group()
print(yt_titles)

#yt_titles = re.search(r'[^|]+',yt_title).group()
print(yt_titles)

The Best of Tchaikovsky | The Most Famous Classical Music Pieces of All Time
The Best of Tchaikovsky | The Most Famous Classical Music Pieces of All Time
The Best of Tchaikovsky 
The Best of Tchaikovsky 


In [31]:
# V0.5.1 --> pass
from pytube import YouTube
import re

yt = YouTube('https://www.youtube.com/watch?v=IgeF8JfotNw')   # 世界名曲（流浪之歌-聖母頌-天鵝-美的羅斯瑪玲-幽默曲-小步舞曲-夢幻曲-愛之喜-布拉姆斯小夜曲-歎息小夜曲-古諾小夜曲-海頓小夜曲-杜立.......)
# yt = YouTube(url) 

yt_audio = yt.streams.filter(abr='128kbps').first()
# yt_audio.download()

yt_title = yt.title
print(yt_title)

yt_titles = re.search(r'[^|]+',yt_title).group()
print(yt_titles)

yt_audio.download(output_path='C:/Users/Lu/Documents/youtube',filename=f'yt_audio_{yt_titles}.mp4', skip_existing=True, max_retries=0)

print(f'downdload < {yt.title} > 完成')

世界名曲（流浪之歌-聖母頌-天鵝-美的羅斯瑪玲-幽默曲-小步舞曲-夢幻曲-愛之喜-布拉姆斯小夜曲-歎息小夜曲-古諾小夜曲-海頓小夜曲-杜立歌小夜曲-小犬小夜曲-杜鵑小夜曲-皇帝圓舞曲-亦然音樂071）
世界名曲（流浪之歌-聖母頌-天鵝-美的羅斯瑪玲-幽默曲-小步舞曲-夢幻曲-愛之喜-布拉姆斯小夜曲-歎息小夜曲-古諾小夜曲-海頓小夜曲-杜立歌小夜曲-小犬小夜曲-杜鵑小夜曲-皇帝圓舞曲-亦然音樂071）
downdload < 世界名曲（流浪之歌-聖母頌-天鵝-美的羅斯瑪玲-幽默曲-小步舞曲-夢幻曲-愛之喜-布拉姆斯小夜曲-歎息小夜曲-古諾小夜曲-海頓小夜曲-杜立歌小夜曲-小犬小夜曲-杜鵑小夜曲-皇帝圓舞曲-亦然音樂071） > 完成



pass               ^^^^^^^^^^^
+++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++
=====================================================================================


In [27]:
# V0.6 --> 2024.06.19 final
# from V0.42 for test_3.1 for window file formate 

from pytube import YouTube
import re

#url = input("Enter the url of media : ")   
url = 'https://www.youtube.com/watch?v=VBNUwMd_eF4&t=4029s'  # 曲目:蕭邦 一定要聽的鋼琴古典樂 ｜放鬆音樂 讀書 專注 The Best "Chopin" Classical Piano Music｜Relaxing Studying Music
yt = YouTube(url)   

#yt_titles = yt.title
print('全名 : ',yt.title)
list=['/',':','"','<','>','\?','\*']
yt_titles = yt.title
for i in list:
    yt_title = re.sub(f'{i}','_',yt_titles)
    yt_titles = yt_title
    # print(i)
print(yt_titles)

yt_title = re.search(r'[A-Za-z0-9_ ]+',yt_titles).group()
# print(yt_titles)
print('簡稱 : ',yt_title)

yt_audio = yt.streams.filter(abr='128kbps').first()

yt_audio.download(output_path='C:/Users/Lu/Documents/youtube',filename=f'yt_audio_{yt_title}.mp4', skip_existing=True, max_retries=0)   #執行ok

print(f'downdload < {yt.title} > 完成')

全名 :  The Best "Chopin" Classical Piano Music｜Relaxing Studying Music
The Best _Chopin_ Classical Piano Music｜Relaxing Studying Music
簡稱 :  The Best _Chopin_ Classical Piano Music
downdload < The Best "Chopin" Classical Piano Music｜Relaxing Studying Music > 完成


In [32]:
# V0.6 --> test of downloading stream

from pytube import YouTube
import re

#url = input("Enter the url of media : ")   
url = 'https://www.youtube.com/watch?v=X3PnA6TUovQ'  # 曲目:最好的古典音樂。靈魂音樂：貝多芬，莫扎特，舒伯特，肖邦，巴赫...🎶🎶卷27
yt = YouTube(url)   

#yt_titles = yt.title
print('全名 : ',yt.title)
list=['/',':','"','<','>','\?','\*']
yt_titles = yt.title
for i in list:
    yt_title = re.sub(f'{i}','_',yt_titles)
    yt_titles = yt_title
    # print(i)
print(yt_titles)

yt_title = re.search(r'[A-Za-z0-9_ ]+',yt_titles).group()
# print(yt_titles)
#print('簡稱 : ',yt_title)

if yt_title == " " :
    yt_title = 'unknown name'
    print('unknown name')
else:
    yt_title = yt_title
print('簡稱 : ',yt_title)

yt_audio = yt.streams.filter(abr='128kbps').first()

#yt_audio.download(output_path='C:/Users/Lu/Documents/youtube',filename=f'yt_audio_{yt_title}.mp4', skip_existing=True, max_retries=0)   #執行ok

print(f'downdload < {yt.title} > 完成')

# 1m 25.2s --> 203,343Kb

全名 :  Лучшая классическая музыка. Музыка для души: Бетховен, Моцарт, Шуберт, Шопен, Бах ... 🎶🎶 Том 27
Лучшая классическая музыка. Музыка для души_ Бетховен, Моцарт, Шуберт, Шопен, Бах ... 🎶🎶 Том 27
unknown name
簡稱 :  unknown name
downdload < Лучшая классическая музыка. Музыка для души: Бетховен, Моцарт, Шуберт, Шопен, Бах ... 🎶🎶 Том 27 > 完成


+++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++
V0.7 pass --> timeout for downloading

In [33]:
# V0.7 --> test of downloading stream for timeout

from pytube import YouTube
import re
import subprocess

#url = input("Enter the url of media : ")   
url = 'https://www.youtube.com/watch?v=X3PnA6TUovQ'  # 曲目:最好的古典音樂。靈魂音樂：貝多芬，莫扎特，舒伯特，肖邦，巴赫...🎶🎶卷27
yt = YouTube(url)   

#yt_titles = yt.title
print('全名 : ',yt.title)
list=['/',':','"','<','>','\?','\*']
yt_titles = yt.title
for i in list:
    yt_title = re.sub(f'{i}','_',yt_titles)
    yt_titles = yt_title
    # print(i)
print(yt_titles)

yt_title = re.search(r'[A-Za-z0-9_ ]+',yt_titles).group()
# print(yt_titles)
#print('簡稱 : ',yt_title)

if yt_title == " " :
    yt_title = 'unknown name'
    print('unknown name')
else:
    yt_title = yt_title
print('簡稱 : ',yt_title)

yt_audio = yt.streams.filter(abr='128kbps').first()



try:
    print("執行download命令")  # 增加執行內容信息
    yt_audio.download(output_path='C:/Users/Lu/Documents/youtube',filename=f'yt_audio_{yt_title}.mp4', skip_existing=True, max_retries=0,timeout=60)
    # 增加timeout=60之pytube.exceptions.PytubeError檢查
    print("成功！輸出文件:")
except yt_audio.exceptions.PytubeError as e:
    print("過程中發生錯誤:", e)
    if e.stderr:
        print("命令輸出:", e.stderr)
    else:
        print("無法獲取 錯誤輸出。")
except subprocess.TimeoutExpired as e:
    print("命令執行超時:", e)


#yt_audio.download(output_path='C:/Users/Lu/Documents/youtube',filename=f'yt_audio_{yt_title}.mp4', skip_existing=True, max_retries=0)   #執行ok

print(f'downdload < {yt.title} > 完成')

# 1m 25.2s --> 203,343Kb

全名 :  Лучшая классическая музыка. Музыка для души: Бетховен, Моцарт, Шуберт, Шопен, Бах ... 🎶🎶 Том 27
Лучшая классическая музыка. Музыка для души_ Бетховен, Моцарт, Шуберт, Шопен, Бах ... 🎶🎶 Том 27
unknown name
簡稱 :  unknown name
執行download命令
成功！輸出文件:
downdload < Лучшая классическая музыка. Музыка для души: Бетховен, Моцарт, Шуберт, Шопен, Бах ... 🎶🎶 Том 27 > 完成


In [1]:
# V0.71 --> test of downloading stream for timeout

from pytube import YouTube
import re
import subprocess
from pytube.exceptions import PytubeError

#url = input("Enter the url of media : ")   
url = 'https://www.youtube.com/watch?v=iD5jUMf0SFM&t=3639s'  # 蕭邦 夜曲｜鋼琴、古典音樂｜放鬆 讀書音樂 The Best Nocturnes by Chopin | Studying, Relaxing Music
yt = YouTube(url)   

# yt.title設定符合 windows 檔案格式
print('全名 : ',yt.title)
list=['/',':','"','<','>','\?','\*']
yt_titles = yt.title
for i in list:
    yt_title = re.sub(f'{i}','_',yt_titles)
    yt_titles = yt_title
print(yt_titles)

yt_title = re.search(r'[A-Za-z0-9_ ]+',yt_titles).group()

if yt_title == " " :
    yt_title = 'unknown name'
    print('unknown name')
else:
    yt_title = yt_title
print('簡稱 : ',yt_title)

yt_audio = yt.streams.filter(abr='128kbps').first()

# 增加timeout=60之pytube.exceptions.PytubeError檢查
try:
    print("執行download命令中.......")  # 增加執行內容信息
    yt_audio.download(output_path='C:/Users/Lu/Documents/youtube',filename=f'yt_audio_{yt_title}.mp4', skip_existing=True, max_retries=0,timeout=60)
    print("成功！輸出文件:")
except PytubeError as e:
    print("過程中發生 pytube 錯誤:", e)
except Exception as e:
    print("發生其他錯誤:", e)

print(f'downdload < {yt.title} > 完成')

print('儲存路徑 < C:/Users/Lu/Documents/youtube > ')
print(f'檔案名稱 : yt_audio_{yt_title}.mp4')

# 45.9s --> 108,035Kb

全名 :  The Best Nocturnes by Chopin | Studying, Relaxing Music
The Best Nocturnes by Chopin | Studying, Relaxing Music
簡稱 :  The Best Nocturnes by Chopin 
執行download命令
成功！輸出文件:
downdload < The Best Nocturnes by Chopin | Studying, Relaxing Music > 完成


====================================================================
++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++
### 設定時間限制(timeout)之討論

#### 透過chatGPT -1
##### Question : 若無預期的錯誤，但很久無法完成，如何印出訊息，跳出執行？
##### Answer:
要處理在執行下載過程中遇到長時間無法完成的情況，可以使用 signal 模組來設定超時處理。在超時發生時，會拋出一個自定義的例外，從而中斷下載過程並打印錯誤訊息。以下是示範程式碼：

In [ ]:
# for iOS ???????
import signal
from pytube import YouTube
from pytube.exceptions import PytubeError

# 自定義的超時例外
class TimeoutException(Exception):
    pass

# 處理超時的函數
def timeout_handler(signum, frame):
    raise TimeoutException()

# 設定超時（例如 60 秒）
TIMEOUT = 60

def download_video(video_url):
    # 在開始下載前設定超時
    signal.signal(signal.SIGALRM, timeout_handler)
    signal.alarm(TIMEOUT)
    
    try:
        video = YouTube(video_url)
        stream = video.streams.get_highest_resolution()
        stream.download()
        print("下載成功！")
    except TimeoutException:
        print("下載過程超時，請重試。")
    except PytubeError as e:
        print("發生 pytube 錯誤:", e)
    except Exception as e:
        print("發生其他錯誤:", e)
    finally:
        # 取消超時
        signal.alarm(0)

if __name__ == "__main__":
    video_url = input("輸入 YouTube 視頻網址: ")
    download_video(video_url)


在這段程式碼中，我們做了以下幾點：

定義了一個自定義的例外 TimeoutException。
使用 signal 模組來設定超時處理函數 timeout_handler。
在開始下載前，設定了一個 60 秒的超時。
在 try 區塊中進行視頻下載操作。如果下載過程中超過 60 秒未完成，則會觸發 TimeoutException 並打印相應的錯誤訊息。
在 finally 區塊中取消超時設定，以防止影響後續操作。
這樣，如果下載過程超過設定的時間限制，程式將中斷執行並打印提示訊息。

#### 透過chatGPT -2
##### Question : stream.download(timeout=60)無法運用嗎？
##### Answer :
pytube 庫中的 download 方法本身並沒有內建的超時機制。因此，無法直接通過設置 timeout 參數來控制下載時間。不過，可以通過一些額外的處理來實現類似的功能。

使用 signal 模組設置超時是比較常見的方法之一，但需要注意的是，這種方法在 Windows 平台上可能會遇到一些問題，因為 signal.alarm 在 Windows 上不受支持。

如果你需要在 Windows 上實現下載超時，可以考慮使用多線程或多進程來達到相似的效果。例如，使用 threading 模組來實現超時控制。以下是一個示例：

In [ ]:
# for windows
import threading
import time
from pytube import YouTube
from pytube.exceptions import PytubeError

# 自定義的超時例外
class TimeoutException(Exception):
    pass

def download_video_with_timeout(video_url, timeout):
    def target():
        try:
            video = YouTube(video_url)
            stream = video.streams.get_highest_resolution()
            stream.download()
            print("下載成功！")
        except PytubeError as e:
            print("發生 pytube 錯誤:", e)
        except Exception as e:
            print("發生其他錯誤:", e)

    thread = threading.Thread(target=target)
    thread.start()
    thread.join(timeout)
    
    if thread.is_alive():
        print("下載過程超時，請重試。")
        raise TimeoutException()

if __name__ == "__main__":
    video_url = input("輸入 YouTube 視頻網址: ")
    timeout = 60  # 設置超時秒數
    try:
        download_video_with_timeout(video_url, timeout)
    except TimeoutException:
        print("下載操作被強制終止。")


在這個示例中：

定義了 download_video_with_timeout 函數，其中啟動了一個新線程來執行下載操作。
使用 thread.join(timeout) 方法等待下載線程完成，並設置超時時間。如果超時，則打印提示訊息並拋出 TimeoutException。
在主程序中，捕獲 TimeoutException 並打印相應的提示訊息。
這種方法在 Windows 和 Unix 系統上都能正常工作，可以有效地設置下載操作的超時時間。

#### 透過chatGPT -3
##### Question : thread = threading.Thread(target=target)語法為何？
##### Answer :
threading.Thread(target=target) 是用來創建一個新的線程（Thread）。在這個語法中：
1. threading.Thread 是 Python 標準庫中的 threading 模組的一個類，用來創建並管理線程。
2. target=target 是指定線程要執行的目標函數。在這個例子中，target 是一個函數名。

以下是對這個語法的詳細解釋：
1. threading.Thread 類：
  threading.Thread 是用來創建一個新的線程對象。每個線程對象都會運行指定的目標函數。
2. target=target：
  target 參數用來指定線程在啟動後要執行的函數。這個函數不接受任何參數。
  在這個例子中，target 是一個函數名，這個函數會在新線程中運行。

以下是一個完整的例子，說明如何使用 threading.Thread 創建並啟動一個新線程來執行指定的函數：

In [2]:
import threading
import time

# 定義一個函數，將在線程中運行
def print_hello():
    for i in range(5):
        print(f"Hello from thread {i}")
        time.sleep(1)  # 每次打印後等待1秒

# 創建一個線程對象，將 `print_hello` 函數作為目標函數
thread = threading.Thread(target=print_hello)

# 啟動線程，開始執行 `print_hello` 函數
thread.start()

# 在主線程中也打印一些信息
for i in range(5):
    print(f"Hello from main thread {i}")
    time.sleep(0.5)  # 每次打印後等待0.5秒

# 等待子線程結束
thread.join()

print("Thread finished.")


Hello from thread 0
Hello from main thread 0
Hello from main thread 1
Hello from thread 1
Hello from main thread 2
Hello from main thread 3
Hello from thread 2
Hello from main thread 4
Hello from thread 3
Hello from thread 4
Thread finished.


在這個例子中：

創建了一個名為 print_hello 的函數，這個函數將在新線程中運行。
使用 threading.Thread(target=print_hello) 創建了一個新的線程對象，並將 print_hello 函數指定為目標函數。
使用 thread.start() 啟動新線程，這會開始執行 print_hello 函數。
在主線程中，打印一些信息來顯示主線程和子線程是如何並行運行的。
使用 thread.join() 等待子線程結束。
這樣，新線程和主線程可以同時運行，這就是多線程編程的基本原理。

#### 透過chatGPT -4
##### Question : thread.join()一定需要嗎？
##### Answer :
thread.join() 在多線程編程中並不是一定需要的，但它在許多情況下是有用的。以下是一些情況以及是否需要 thread.join() 的原因：

##### 何時需要使用 thread.join()：
1.等待線程完成：
  如果你需要確保某個線程完成它的工作後，才能繼續執行主程序中的後續步驟，那麼你應該使用 thread.join()。
例如，如果你有一個線程負責下載文件，而主程序需要處理這些文件，那麼你需要確保下載完成後再進行處理。這時可以使用 thread.join()。
2.同步多個線程：
  當你有多個線程，並且需要等待所有線程完成後再進行下一步操作，可以對每個線程調用 join()。
這樣可以確保主程序在所有線程結束後再繼續執行。

##### 何時不需要使用 thread.join()：
1.非阻塞執行：
  如果你希望主程序和線程並行執行，而不需要等待線程完成，那麼可以不使用 thread.join()。
例如，在 GUI 程序中，你可能希望在背景線程中執行長時間運行的任務，而不阻塞主線程的用戶界面更新。
2.獨立任務：
  如果線程執行的任務與主程序完全獨立，沒有相互依賴，也可以不使用 thread.join()。
例如，記錄日誌的線程可以獨立運行，無需等待它結束。

示例代碼
這裡有兩個示例，分別說明使用和不使用 thread.join() 的情況：

使用 thread.join()

In [3]:
# 使用 thread.join()
import threading
import time

def worker():
    print("Worker thread started")
    time.sleep(2)
    print("Worker thread finished")

# 創建並啟動線程
thread = threading.Thread(target=worker)
thread.start()

# 等待線程結束
thread.join()
print("Main thread finished after worker thread")


Worker thread started
Worker thread finished
Main thread finished after worker thread


在這個示例中，主程序會等待 worker 線程完成後再繼續執行，保證 Worker thread finished 先於 Main thread finished after worker thread 被打印。

不使用 thread.join()

In [4]:
# 不使用 thread.join()
import threading
import time

def worker():
    print("Worker thread started")
    time.sleep(2)
    print("Worker thread finished")

# 創建並啟動線程
thread = threading.Thread(target=worker)
thread.start()

# 主程序不等待線程結束
print("Main thread finished immediately")


Worker thread started
Main thread finished immediately


Worker thread finished


在這個示例中，主程序不等待 worker 線程完成，會立即打印 Main thread finished immediately，然後 worker 線程會在其自己的時間內完成。

總結來說，是否使用 thread.join() 取決於具體需求。如果需要等待線程完成才能繼續下一步操作，那麼應該使用 join()；如果不需要等待，則可以不使用。

### 設定時間限制(timeout)之討論
    ~~~~~   End  ~~~~~
A+++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++
B===================================================================

====================================================================
++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++
### 使用FFmpeg將音頻MP4轉換為MP3?之討論

#### 透過chatGPT -1
##### Question :使用FFmpeg將音頻MP4轉換為MP3?
##### Answer:
要處理在執行下載過程中遇到長時間無法完成的情況，可以使用 signal 模組來設定超時處理。在超時發生時，會拋出一個自定義的例外，從而中斷下載過程並打印錯誤訊息。以下是示範程式碼：
要使用 FFmpeg 將音頻從 MP4 轉換為 MP3，您可以使用以下命令：

>>>ffmpeg -i input.mp4 -vn -acodec libmp3lame -q:a 2 output.mp3

這裡是這個命令的詳細說明：
1. ffmpeg: 調用 FFmpeg 工具。
2. -i input.mp4: 指定輸入文件 input.mp4。
3. -vn: 禁用視頻，因為我們只需要音頻部分。
4. -acodec libmp3lame: 指定 MP3 編碼器。libmp3lame 是一個高質量的 MP3 編碼器。
5. -q:a 2: 設置音頻質量。q:a 選項允許設置 VBR（可變比特率）的質量級別，數字 2 表示非常高的質量（範圍從 0 到 9，0 為最佳）。
6. output.mp3: 指定輸出文件 output.mp3。

您可以將此命令用於命令行中，也可以在 Python 程序中通過 subprocess 模組調用它。這裡有一個示例 Python 程序，演示如何使用 subprocess 來運行 FFmpeg 命令：

In [ ]:
import subprocess

def convert_mp4_to_mp3(input_file, output_file):
    command = [
        'ffmpeg',
        '-i', input_file,
        '-vn',
        '-acodec', 'libmp3lame',
        '-q:a', '2',
        output_file
    ]
    
    try:
        result = subprocess.run(command, check=True, stdout=subprocess.PIPE, stderr=subprocess.PIPE, text=True)
        print(f"Conversion successful! MP3 file is located at: {output_file}")
    except subprocess.CalledProcessError as e:
        print("An error occurred during conversion:", e)
        if e.stderr:
            print("FFmpeg error output:", e.stderr)
        else:
            print("Could not retrieve FFmpeg error output.")

# Example usage
convert_mp4_to_mp3("input.mp4", "output.mp3")


這個程序調用 convert_mp4_to_mp3 函數，將 input.mp4 轉換為 output.mp3。如果轉換成功，它會打印成功消息；如果發生錯誤，則會打印錯誤消息和 FFmpeg 的錯誤輸出。

#### 透過chatGPT -2
##### Question : 當abr=128kbps 時，-q:a 要設置為何?
##### Answer:
在FFmpeg中，-q:a 和 -b:a 是用來設定音頻質量和比特率的參數。這兩者通常不會同時使用。-q:a（質量）用來控制VBR（可變比特率）的音質，數值範圍從0到9，數值越小音質越好。而 -b:a 用來設定CBR（固定比特率）的音頻比特率。

如果你想要設定音頻比特率為128kbps，應該使用 -b:a 而不是 -q:a。

以下是使用 -b:a 的正確方式

In [ ]:
def convert_mp4_to_mp3(input_file, output_file):
    command = [
        'ffmpeg',
        '-i', input_file,
        '-vn',
        '-acodec', 'libmp3lame',
        '-b:a', '128k',
        output_file
    ]
    try:
        result = subprocess.run(command, check=True, stdout=subprocess.PIPE, stderr=subprocess.PIPE, text=True)
        print("轉換成功！輸出文件位於:", output_file)
    except subprocess.CalledProcessError as e:
        print("ffmpeg轉換過程中發生錯誤:", e)
        if e.stderr:
            print("ffmpeg命令輸出:", e.stderr)
        else:
            print("無法獲取 ffmpeg 錯誤輸出。")

# 範例使用
convert_mp4_to_mp3("input.mp4", "output.mp3")


這樣可以確保音頻的比特率是128kbps。如果你還是想要使用 -q:a 來設定音質的話，你需要注意，-q:a 和比特率之間沒有一個簡單的直接對應，通常需要實驗和經驗來找出適合你的設置。對應 -q:a 設置的比特率值大致如下：

-q:a 0 大約是 320 kbps
-q:a 1 大約是 256 kbps
-q:a 2 大約是 192 kbps
-q:a 3 大約是 160 kbps
-q:a 4 大約是 128 kbps
-q:a 5 大約是 96 kbps
所以如果你希望音質接近128kbps，可以設定 -q:a 4。但請注意這只是近似值。

In [ ]:
def convert_mp4_to_mp3(input_file, output_file):
    command = [
        'ffmpeg',
        '-i', input_file,
        '-vn',
        '-acodec', 'libmp3lame',
        '-q:a', '4',
        output_file
    ]
    try:
        result = subprocess.run(command, check=True, stdout=subprocess.PIPE, stderr=subprocess.PIPE, text=True)
        print("轉換成功！輸出文件位於:", output_file)
    except subprocess.CalledProcessError as e:
        print("ffmpeg轉換過程中發生錯誤:", e)
        if e.stderr:
            print("ffmpeg命令輸出:", e.stderr)
        else:
            print("無法獲取 ffmpeg 錯誤輸出。")

# 範例使用
convert_mp4_to_mp3("input.mp4", "output.mp3")


       ~~~ End ~~~
+++++++++++++++++++++++++++++
=============================

In [1]:
# mp4 --> mp3 轉換確認

import subprocess

def convert_mp4_to_mp3(input_file, output_file):
    command = [
        'ffmpeg',
        '-i', input_file,
        '-vn',
        '-acodec', 'libmp3lame',
        '-q:a', '3',
        output_file
    ]
    # -q:a 3 大約是 160 kbps
    try:
        result = subprocess.run(command, check=True, stdout=subprocess.PIPE, stderr=subprocess.PIPE, text=True)
        print("轉換成功！輸出文件位於:", output_file)
    except subprocess.CalledProcessError as e:
        print("ffmpeg轉換過程中發生錯誤:", e)
        if e.stderr:
            print("ffmpeg命令輸出:", e.stderr)
        else:
            print("無法獲取 ffmpeg 錯誤輸出。")

# 範例使用
# convert_mp4_to_mp3("input.mp4", "output.mp3")
convert_mp4_to_mp3(f'C:/Users/Lu/Documents/youtube/yt_audio_ Best Classical Music.mp4', f'C:/Users/Lu/Documents/youtube/yt_audio_ Best Classical Music.mp3')
# media 長度 03:34:26 mp4容量大小203,343KB mp3容量大小225,418KB 處理耗時4m 10.1s

轉換成功！輸出文件位於: C:/Users/Lu/Documents/youtube/yt_audio_ Best Classical Music.mp3


In [3]:
# V0.80
# get_audio_only V0.6 --> 2024.06.19 final

from pytube import YouTube
import re
import subprocess

#url = input("Enter the url of media : ")   
#url = 'https://www.youtube.com/watch?v=VBNUwMd_eF4&t=4029s'  # 曲目:蕭邦 一定要聽的鋼琴古典樂 ｜放鬆音樂 讀書 專注 The Best "Chopin" Classical Piano Music｜Relaxing Studying Music

def get_audio_only(url):
    yt = YouTube(url)   
    print('全名 : ',yt.title)
    list=['/',':','"','<','>','\?','\*']
    yt_titles = yt.title
    for i in list:
        yt_title = re.sub(f'{i}','_',yt_titles)
        yt_titles = yt_title
    print(yt_titles)

    yt_title = re.search(r'[A-Za-z0-9_ ]+',yt_titles).group()
    print('簡稱 : ',yt_title)

    yt_audio = yt.streams.filter(abr='128kbps').first()
    yt_audio.download(output_path='C:/Users/Lu/Documents/youtube',filename=f'yt_audio_{yt_title}.mp4', skip_existing=True, max_retries=0)   #執行ok
    print(f'downdload < {yt.title} > 完成')
    return yt_title

def convert_mp4_to_mp3(input_file, output_file):
    command = [
        'ffmpeg',
        '-i', input_file,
        '-vn',
        '-acodec', 'libmp3lame',
        '-q:a', '3',
        output_file
    ]
    # -q:a 3 大約是 160 kbps
    try:
        result = subprocess.run(command, check=True, stdout=subprocess.PIPE, stderr=subprocess.PIPE, text=True)
        print("轉換成功！輸出文件位於:", output_file)
    except subprocess.CalledProcessError as e:
        print("ffmpeg轉換過程中發生錯誤:", e)
        if e.stderr:
            print("ffmpeg命令輸出:", e.stderr)
        else:
            print("無法獲取 ffmpeg 錯誤輸出。")


url = 'https://www.youtube.com/watch?v=JmxrMBuCsoI&t=121s'  # 曲目:The Beautiful Blue Danube - Johann Strauss (Squid Game) BGM 小約翰 施特勞斯 藍色多瑙河 魷魚遊戲 Op.314
get_audio_only(url)
# convert_mp4_to_mp3("input.mp4", "output.mp3")# 範例使用
yt_title = get_audio_only(url)  #返回結果方式
convert_mp4_to_mp3(f'C:/Users/Lu/Documents/youtube/yt_audio_{yt_title}.mp4', f'C:/Users/Lu/Documents/youtube/yt_audio_{yt_title}.mp3')

# media 長度 01:08:39 mp4容量大小65,114KB mp3容量大小73,261KB 處理耗時2m 37.6s

全名 :  The Beautiful Blue Danube - Johann Strauss (Squid Game) BGM 小約翰 施特勞斯 藍色多瑙河 魷魚遊戲 Op.314
The Beautiful Blue Danube - Johann Strauss (Squid Game) BGM 小約翰 施特勞斯 藍色多瑙河 魷魚遊戲 Op.314
簡稱 :  The Beautiful Blue Danube 
downdload < The Beautiful Blue Danube - Johann Strauss (Squid Game) BGM 小約翰 施特勞斯 藍色多瑙河 魷魚遊戲 Op.314 > 完成
全名 :  The Beautiful Blue Danube - Johann Strauss (Squid Game) BGM 小約翰 施特勞斯 藍色多瑙河 魷魚遊戲 Op.314
The Beautiful Blue Danube - Johann Strauss (Squid Game) BGM 小約翰 施特勞斯 藍色多瑙河 魷魚遊戲 Op.314
簡稱 :  The Beautiful Blue Danube 
downdload < The Beautiful Blue Danube - Johann Strauss (Squid Game) BGM 小約翰 施特勞斯 藍色多瑙河 魷魚遊戲 Op.314 > 完成
轉換成功！輸出文件位於: C:/Users/Lu/Documents/youtube/yt_audio_The Beautiful Blue Danube .mp3


In [4]:
# V0.81
# get_audio_only V0.6 --> 2024.06.19 final
# 定義全局變量 global

from pytube import YouTube
import re
import subprocess

#url = input("Enter the url of media : ")   
#url = 'https://www.youtube.com/watch?v=VBNUwMd_eF4&t=4029s'  # 曲目:蕭邦 一定要聽的鋼琴古典樂 ｜放鬆音樂 讀書 專注 The Best "Chopin" Classical Piano Music｜Relaxing Studying Music

# 定義全局變量
global_yt_title = None

def get_audio_only(url):
    yt = YouTube(url)   
    print('全名 : ',yt.title)
    list=['/',':','"','<','>','\?','\*']
    yt_titles = yt.title
    for i in list:
        yt_title = re.sub(f'{i}','_',yt_titles)
        yt_titles = yt_title
    print(yt_titles)

    yt_title = re.search(r'[A-Za-z0-9_ ]+',yt_titles).group()
    print('簡稱 : ',yt_title)
    global global_yt_title   # 定義全局變量
    global_yt_title = yt_title    # 定義全局變量

    yt_audio = yt.streams.filter(abr='128kbps').first()
    yt_audio.download(output_path='C:/Users/Lu/Documents/youtube',filename=f'yt_audio_{yt_title}.mp4', skip_existing=True, max_retries=0)   #執行ok
    print(f'downdload < {yt.title} > 完成')
    return yt_title

def convert_mp4_to_mp3(input_file, output_file):
    command = [
        'ffmpeg',
        '-i', input_file,
        '-vn',
        '-acodec', 'libmp3lame',
        '-q:a', '3',
        output_file
    ]
    # -q:a 3 大約是 160 kbps
    try:
        result = subprocess.run(command, check=True, stdout=subprocess.PIPE, stderr=subprocess.PIPE, text=True)
        print("轉換成功！輸出文件位於:", output_file)
    except subprocess.CalledProcessError as e:
        print("ffmpeg轉換過程中發生錯誤:", e)
        if e.stderr:
            print("ffmpeg命令輸出:", e.stderr)
        else:
            print("無法獲取 ffmpeg 錯誤輸出。")


url = 'https://www.youtube.com/watch?v=JmxrMBuCsoI&t=121s'  # 曲目:The Beautiful Blue Danube - Johann Strauss (Squid Game) BGM 小約翰 施特勞斯 藍色多瑙河 魷魚遊戲 Op.314
get_audio_only(url)
# convert_mp4_to_mp3("input.mp4", "output.mp3")# 範例使用
# yt_title = get_audio_only(url)  #返回結果方式

yt_title = global_yt_title     # 定義全局變量

convert_mp4_to_mp3(f'C:/Users/Lu/Documents/youtube/yt_audio_{yt_title}.mp4', f'C:/Users/Lu/Documents/youtube/yt_audio_{yt_title}.mp3')

# media 長度 01:08:39 mp4容量大小65,114KB mp3容量大小73,261KB 處理耗時2m 30.1s

全名 :  The Beautiful Blue Danube - Johann Strauss (Squid Game) BGM 小約翰 施特勞斯 藍色多瑙河 魷魚遊戲 Op.314
The Beautiful Blue Danube - Johann Strauss (Squid Game) BGM 小約翰 施特勞斯 藍色多瑙河 魷魚遊戲 Op.314
簡稱 :  The Beautiful Blue Danube 
downdload < The Beautiful Blue Danube - Johann Strauss (Squid Game) BGM 小約翰 施特勞斯 藍色多瑙河 魷魚遊戲 Op.314 > 完成
轉換成功！輸出文件位於: C:/Users/Lu/Documents/youtube/yt_audio_The Beautiful Blue Danube .mp3


In [5]:
# V0.81 驗證01
# get_audio_only V0.6 --> 2024.06.19 final
# 定義全局變量 global

from pytube import YouTube
import re
import subprocess

#url = input("Enter the url of media : ")   
#url = 'https://www.youtube.com/watch?v=VBNUwMd_eF4&t=4029s'  # 曲目:蕭邦 一定要聽的鋼琴古典樂 ｜放鬆音樂 讀書 專注 The Best "Chopin" Classical Piano Music｜Relaxing Studying Music

# 定義全局變量
global_yt_title = None

def get_audio_only(url):
    yt = YouTube(url)   
    print('全名 : ',yt.title)
    list=['/',':','"','<','>','\?','\*']
    yt_titles = yt.title
    for i in list:
        yt_title = re.sub(f'{i}','_',yt_titles)
        yt_titles = yt_title
    print(yt_titles)

    yt_title = re.search(r'[A-Za-z0-9_ ]+',yt_titles).group()
    print('簡稱 : ',yt_title)
    global global_yt_title   # 定義全局變量
    global_yt_title = yt_title    # 定義全局變量

    yt_audio = yt.streams.filter(abr='128kbps').first()
    yt_audio.download(output_path='C:/Users/Lu/Documents/youtube',filename=f'yt_audio_{yt_title}.mp4', skip_existing=True, max_retries=0)   #執行ok
    print(f'downdload < {yt.title} > 完成')
    return yt_title

def convert_mp4_to_mp3(input_file, output_file):
    command = [
        'ffmpeg',
        '-i', input_file,
        '-vn',
        '-acodec', 'libmp3lame',
        '-q:a', '3',
        output_file
    ]
    # -q:a 3 大約是 160 kbps
    try:
        result = subprocess.run(command, check=True, stdout=subprocess.PIPE, stderr=subprocess.PIPE, text=True)
        print("轉換成功！輸出文件位於:", output_file)
    except subprocess.CalledProcessError as e:
        print("ffmpeg轉換過程中發生錯誤:", e)
        if e.stderr:
            print("ffmpeg命令輸出:", e.stderr)
        else:
            print("無法獲取 ffmpeg 錯誤輸出。")


url = 'https://www.youtube.com/watch?v=oYQut39Uyww'  # 曲目:2021宮崎駿鋼琴音樂合集-最佳放鬆音樂《100%無廣告》スタジオジブリGhibli Relaxing Music【スターバックス】工作-讀書專用音樂
get_audio_only(url)
# convert_mp4_to_mp3("input.mp4", "output.mp3")# 範例使用
# yt_title = get_audio_only(url)  #返回結果方式

yt_title = global_yt_title     # 定義全局變量

convert_mp4_to_mp3(f'C:/Users/Lu/Documents/youtube/yt_audio_{yt_title}.mp4', f'C:/Users/Lu/Documents/youtube/yt_audio_{yt_title}.mp3')

# media 長度 01:33:15 mp4容量大小88,433KB mp3容量大小87,074KB 處理耗時2m 2.9s

全名 :  2021宮崎駿鋼琴音樂合集-最佳放鬆音樂《100%無廣告》スタジオジブリGhibli Relaxing Music【スターバックス】工作-讀書專用音樂
2021宮崎駿鋼琴音樂合集-最佳放鬆音樂《100%無廣告》スタジオジブリGhibli Relaxing Music【スターバックス】工作-讀書專用音樂
簡稱 :  2021
downdload < 2021宮崎駿鋼琴音樂合集-最佳放鬆音樂《100%無廣告》スタジオジブリGhibli Relaxing Music【スターバックス】工作-讀書專用音樂 > 完成
轉換成功！輸出文件位於: C:/Users/Lu/Documents/youtube/yt_audio_2021.mp3


解釋
\w 匹配字母、數字和下劃線，即 [a-zA-Z0-9_]
\u4e00-\u9fff 匹配 Unicode 范圍內的中文字符
[\w\u4e00-\u9fff]+ 匹配一個或多個中英文字符
如果你想要更精確地匹配，可能需要根據實際需求進行調整。例如，如果你只想匹配中英文字符而不包括數字，可以使用 [a-zA-Z\u4e00-\u9fff]+。

更多示例
只匹配中英文字符（不包括數字和下劃線）：
>>>pattern = r'[a-zA-Z\u4e00-\u9fff]+'

匹配中英文字符及空格（允許空格分隔的詞組）：
>>>pattern = r'[a-zA-Z\u4e00-\u9fff\s]+'

匹配中英文字符及標點符號：
>>>pattern = r'[a-zA-Z\u4e00-\u9fff.,!?，。！？]+'

使用這些模式可以靈活匹配各種包含中英文字符的字符串。根據具體需求選擇合適的正則表達式模式。

In [7]:
# V0.81 驗證02
# get_audio_only V0.6 --> 2024.06.19 final
# 定義全局變量 global
#設定包含中英文字

from pytube import YouTube
import re
import subprocess

#url = input("Enter the url of media : ")   
#url = 'https://www.youtube.com/watch?v=VBNUwMd_eF4&t=4029s'  # 曲目:蕭邦 一定要聽的鋼琴古典樂 ｜放鬆音樂 讀書 專注 The Best "Chopin" Classical Piano Music｜Relaxing Studying Music

# 定義全局變量
global_yt_title = None

def get_audio_only(url):
    yt = YouTube(url)   
    print('全名 : ',yt.title)
    list=['/',':','"','<','>','\?','\*']
    yt_titles = yt.title
    for i in list:
        yt_title = re.sub(f'{i}','_',yt_titles)
        yt_titles = yt_title
    print(yt_titles)

    #yt_title = re.search(r'[A-Za-z0-9_ \u4e00-\u9fff\s]+',yt_titles).group() #設定包含中英文字
    yt_title = re.search(r'[\w\u4e00-\u9fff\s]+',yt_titles).group() #設定包含中英文字
    print('簡稱 : ',yt_title)
    global global_yt_title   # 定義全局變量
    global_yt_title = yt_title    # 定義全局變量

    yt_audio = yt.streams.filter(abr='128kbps').first()
    yt_audio.download(output_path='C:/Users/Lu/Documents/youtube',filename=f'yt_audio_{yt_title}.mp4', skip_existing=True, max_retries=0)   #執行ok
    print(f'downdload < {yt.title} > 完成')
    return yt_title

def convert_mp4_to_mp3(input_file, output_file):
    command = [
        'ffmpeg',
        '-i', input_file,
        '-vn',
        '-acodec', 'libmp3lame',
        '-q:a', '3',
        output_file
    ]
    # -q:a 3 大約是 160 kbps
    try:
        result = subprocess.run(command, check=True, stdout=subprocess.PIPE, stderr=subprocess.PIPE, text=True, encoding='utf-8')   #追加, encoding='utf-8'
        print("轉換成功！輸出文件位於:", output_file)
    except subprocess.CalledProcessError as e:
        print("ffmpeg轉換過程中發生錯誤:", e)
        if e.stderr:
            print("ffmpeg命令輸出:", e.stderr)
        else:
            print("無法獲取 ffmpeg 錯誤輸出。")


url = 'https://www.youtube.com/watch?v=b5d5OmmUlPc'  # 曲目:2023宮崎駿鋼琴音樂合集-最佳放鬆音樂《100%無廣告》スタジオジブリGhibli Relaxing Music【スターバックス】工作-讀書專用音樂
get_audio_only(url)
# convert_mp4_to_mp3("input.mp4", "output.mp3")# 範例使用
#yt_title = get_audio_only(url)

yt_title = global_yt_title     # 定義全局變量

convert_mp4_to_mp3(f'C:/Users/Lu/Documents/youtube/yt_audio_{yt_title}.mp4', f'C:/Users/Lu/Documents/youtube/yt_audio_{yt_title}.mp3')

# media 長度 01:33:15 mp4容量大小88,433KB mp3容量大小87,074KB 處理耗時2m 2.9s

全名 :  2023宮崎駿鋼琴音樂合集-最佳放鬆音樂《100%無廣告》スタジオジブリGhibli Relaxing Music【スターバックス】工作-讀書專用音樂
2023宮崎駿鋼琴音樂合集-最佳放鬆音樂《100%無廣告》スタジオジブリGhibli Relaxing Music【スターバックス】工作-讀書專用音樂
簡稱 :  2023宮崎駿鋼琴音樂合集
downdload < 2023宮崎駿鋼琴音樂合集-最佳放鬆音樂《100%無廣告》スタジオジブリGhibli Relaxing Music【スターバックス】工作-讀書專用音樂 > 完成
轉換成功！輸出文件位於: C:/Users/Lu/Documents/youtube/yt_audio_2023宮崎駿鋼琴音樂合集.mp3


In [2]:
# V0.82 -->2024.06.22 ok
# get_audio_only V0.6 --> 2024.06.19 final
# 定義全局變量 global
# 設定包含中英文字及空格 r'[\w\u4e00-\u9fff\s]+'

from pytube import YouTube
import re
import subprocess

#url = input("Enter the url of media : ")   
#url = 'https://www.youtube.com/watch?v=VBNUwMd_eF4&t=4029s'  # 曲目:蕭邦 一定要聽的鋼琴古典樂 ｜放鬆音樂 讀書 專注 The Best "Chopin" Classical Piano Music｜Relaxing Studying Music

# 定義全局變量
global_yt_title = None

def get_audio_only(url):
    yt = YouTube(url)   
    print('全名 : ',yt.title)
    list=['/',':','"','<','>','\?','\*']
    yt_titles = yt.title
    for i in list:
        yt_title = re.sub(f'{i}','_',yt_titles)
        yt_titles = yt_title
    print(yt_titles)

    #yt_title = re.search(r'[A-Za-z0-9_ \u4e00-\u9fff\s]+',yt_titles).group() #設定包含中英文字
    yt_title = re.search(r'[\w\u4e00-\u9fff\s]+',yt_titles).group() #設定包含中英文字
    print('簡稱 : ',yt_title)
    global global_yt_title   # 定義全局變量
    global_yt_title = yt_title    # 定義全局變量

    yt_audio = yt.streams.filter(abr='128kbps').first()
    yt_audio.download(output_path='C:/Users/Lu/Documents/youtube',filename=f'yt_audio_{yt_title}.mp4', skip_existing=True, max_retries=0)   #執行ok
    print(f'downdload < {yt.title} > 完成')
    return yt_title

def convert_mp4_to_mp3(input_file, output_file):
    command = [
        'ffmpeg',
        '-i', input_file,
        '-vn',
        '-acodec', 'libmp3lame',
        '-q:a', '3',
        output_file
    ]
    # -q:a 3 大約是 160 kbps
    try:
        result = subprocess.run(command, check=True, stdout=subprocess.PIPE, stderr=subprocess.PIPE, text=True, encoding='utf-8')   #追加, encoding='utf-8'
        print("轉換成功！輸出文件位於:", output_file)
    except subprocess.CalledProcessError as e:
        print("ffmpeg轉換過程中發生錯誤:", e)
        if e.stderr:
            print("ffmpeg命令輸出:", e.stderr)
        else:
            print("無法獲取 ffmpeg 錯誤輸出。")


url = 'https://www.youtube.com/watch?v=7QnAX12M24w'  # 曲目:古典音樂放鬆，安靜的音樂 - 莫扎特，貝多芬，肖邦，Tchaikovsky，Rachmaninov
get_audio_only(url)
# convert_mp4_to_mp3("input.mp4", "output.mp3")# 範例使用
# yt_title = get_audio_only(url)  #返回結果方式

yt_title = global_yt_title     # 定義全局變量

convert_mp4_to_mp3(f'C:/Users/Lu/Documents/youtube/yt_audio_{yt_title}.mp4', f'C:/Users/Lu/Documents/youtube/yt_audio_{yt_title}.mp3')

# media 長度 01:15:53 mp4容量大小71,971KB mp3容量大小71,013KB 處理耗時2m 7.4s

全名 :  Классическая музыка расслабляет, тихая музыкаг- Моцарт, Бетховен, Шопен, Чайковский, Рахманинов
Классическая музыка расслабляет, тихая музыкаг- Моцарт, Бетховен, Шопен, Чайковский, Рахманинов
簡稱 :  Классическая музыка расслабляет
downdload < Классическая музыка расслабляет, тихая музыкаг- Моцарт, Бетховен, Шопен, Чайковский, Рахманинов > 完成
轉換成功！輸出文件位於: C:/Users/Lu/Documents/youtube/yt_audio_Классическая музыка расслабляет.mp3


In [1]:
# V0.82 之應用確認-1

from pytube import YouTube
import re
import subprocess

# 定義全局變量
global_yt_title = None

def get_audio_only(url):
    yt = YouTube(url)   
    print('全名 : ',yt.title)
    list=['/',':','"','<','>','\?','\*']
    yt_titles = yt.title
    for i in list:
        yt_title = re.sub(f'{i}','_',yt_titles)
        yt_titles = yt_title
    print(yt_titles)

    yt_title = re.search(r'[\w\u4e00-\u9fff\s]+',yt_titles).group() #設定包含中英文字
    print('簡稱 : ',yt_title)
    global global_yt_title   # 定義全局變量
    global_yt_title = yt_title    # 定義全局變量

    yt_audio = yt.streams.filter(abr='128kbps').first()
    yt_audio.download(output_path='C:/Users/Lu/Documents/youtube',filename=f'yt_audio_{yt_title}.mp4', skip_existing=True, max_retries=0)   #執行ok
    print(f'downdload < {yt.title} > 完成')
    return yt_title

def convert_mp4_to_mp3(input_file, output_file):
    command = [
        'ffmpeg',
        '-i', input_file,
        '-vn',
        '-acodec', 'libmp3lame',
        '-q:a', '3',
        output_file
    ]
    try:
        result = subprocess.run(command, check=True, stdout=subprocess.PIPE, stderr=subprocess.PIPE, text=True, encoding='utf-8')   #追加, encoding='utf-8'
        print("轉換成功！輸出文件位於:", output_file)
    except subprocess.CalledProcessError as e:
        print("ffmpeg轉換過程中發生錯誤:", e)
        if e.stderr:
            print("ffmpeg命令輸出:", e.stderr)
        else:
            print("無法獲取 ffmpeg 錯誤輸出。")


url = 'https://www.youtube.com/watch?v=BoQq8Zk22pg'  # 曲目:最好的古典音樂。靈魂音樂：貝多芬，莫扎特，舒伯特，肖邦，巴赫。
get_audio_only(url)

yt_title = global_yt_title     # 定義全局變量

convert_mp4_to_mp3(f'C:/Users/Lu/Documents/youtube/yt_audio_{yt_title}.mp4', f'C:/Users/Lu/Documents/youtube/yt_audio_{yt_title}.mp3')

全名 :  Лучшая классическая музыка. Музыка для души: Бетховен, Моцарт, Шуберт, Шопен, Бах .. Том 174 🎧🎧
Лучшая классическая музыка. Музыка для души_ Бетховен, Моцарт, Шуберт, Шопен, Бах .. Том 174 🎧🎧
簡稱 :  Лучшая классическая музыка
downdload < Лучшая классическая музыка. Музыка для души: Бетховен, Моцарт, Шуберт, Шопен, Бах .. Том 174 🎧🎧 > 完成
轉換成功！輸出文件位於: C:/Users/Lu/Documents/youtube/yt_audio_Лучшая классическая музыка.mp3


url = 'https://www.youtube.com/watch?v=BoQq8Zk22pg'  # 曲目:最好的古典音樂。靈魂音樂：貝多芬，莫扎特，舒伯特，肖邦，巴赫。
====== message ==========
全名 :  Лучшая классическая музыка. Музыка для души: Бетховен, Моцарт, Шуберт, Шопен, Бах .. Том 174 🎧🎧
Лучшая классическая музыка. Музыка для души_ Бетховен, Моцарт, Шуберт, Шопен, Бах .. Том 174 🎧🎧
簡稱 :  Лучшая классическая музыка
downdload < Лучшая классическая музыка. Музыка для души: Бетховен, Моцарт, Шуберт, Шопен, Бах .. Том 174 🎧🎧 > 完成
轉換成功！輸出文件位於: C:/Users/Lu/Documents/youtube/yt_audio_Лучшая классическая музыка.mp3

++=++翻譯
The best classical music. Music for the soul: Beethoven, Mozart, Schubert, Chopin, Bach.. Volume 174 🎧🎧
The best classical music. Music for the soul_ Beethoven, Mozart, Schubert, Chopin, Bach .. Volume 174 🎧🎧
簡稱: Best classical music
====== message End==========

#####<ref> media 長度 01:15:53 mp4容量大小71,971KB mp3容量大小71,013KB 處理耗時2m 7.4s
##### media 長度 03:57:21 mp4容量大小225,089KB mp3容量大小236,687KB 處理耗時7m 18.4s

In [2]:
# V0.84 --> 2024.06.23 final
# get_audio_only V0.6 --> 2024.06.19 final
# 以yt_title = get_audio_only(url)返回結果方式(return),取代get_audio_only(url)重複執行

from pytube import YouTube
import re
import subprocess

#url = input("Enter the url of media : ")   
#url = 'https://www.youtube.com/watch?v=VBNUwMd_eF4&t=4029s'  # 曲目:蕭邦 一定要聽的鋼琴古典樂 ｜放鬆音樂 讀書 專注 The Best "Chopin" Classical Piano Music｜Relaxing Studying Music

def get_audio_only(url):
    yt = YouTube(url)   
    print('全名 : ',yt.title)
    list=['/',':','"','<','>','\?','\*']
    yt_titles = yt.title
    for i in list:
        yt_title = re.sub(f'{i}','_',yt_titles)
        yt_titles = yt_title
    print(yt_titles)

    yt_title = re.search(r'[A-Za-z0-9_ ]+',yt_titles).group()
    print('簡稱 : ',yt_title)

    yt_audio = yt.streams.filter(abr='128kbps').first()
    yt_audio.download(output_path='C:/Users/Lu/Documents/youtube',filename=f'yt_audio_{yt_title}.mp4', skip_existing=True, max_retries=0)   #執行ok
    print(f'downdload < {yt.title} > 完成')
    return yt_title

def convert_mp4_to_mp3(input_file, output_file):
    command = [
        'ffmpeg',
        '-i', input_file,
        '-vn',
        '-acodec', 'libmp3lame',
        '-q:a', '3',
        output_file
    ]
    # -q:a 3 大約是 160 kbps
    try:
        result = subprocess.run(command, check=True, stdout=subprocess.PIPE, stderr=subprocess.PIPE, text=True)
        print("轉換成功！輸出文件位於:", output_file)
    except subprocess.CalledProcessError as e:
        print("ffmpeg轉換過程中發生錯誤:", e)
        if e.stderr:
            print("ffmpeg命令輸出:", e.stderr)
        else:
            print("無法獲取 ffmpeg 錯誤輸出。")


url = 'https://www.youtube.com/watch?v=JmxrMBuCsoI&t=121s'  # 曲目:The Beautiful Blue Danube - Johann Strauss (Squid Game) BGM 小約翰 施特勞斯 藍色多瑙河 魷魚遊戲 Op.314
# get_audio_only(url)
yt_title = get_audio_only(url)  #返回結果方式

convert_mp4_to_mp3(f'C:/Users/Lu/Documents/youtube/yt_audio_{yt_title}.mp4', f'C:/Users/Lu/Documents/youtube/yt_audio_{yt_title}.mp3')

# <比較說明>
# media 長度 01:08:39 mp4容量大小65,114KB mp3容量大小73,261KB 處理耗時2m 37.6s-->以get_audio_only(url) & yt_title = get_audio_only(url)重複執行，以取出yt_title給convert_mp4_to_mp3 (v0.80)

# media 長度 01:08:39 mp4容量大小65,114KB mp3容量大小73,261KB 處理耗時2m 30.1s-->使用global全局變量(v0.81 & v0.82)

# media 長度 01:08:39 mp4容量大小65,114KB mp3容量大小73,261KB 處理耗時1m 54.9s-->v0.84

全名 :  The Beautiful Blue Danube - Johann Strauss (Squid Game) BGM 小約翰 施特勞斯 藍色多瑙河 魷魚遊戲 Op.314
The Beautiful Blue Danube - Johann Strauss (Squid Game) BGM 小約翰 施特勞斯 藍色多瑙河 魷魚遊戲 Op.314
簡稱 :  The Beautiful Blue Danube 
downdload < The Beautiful Blue Danube - Johann Strauss (Squid Game) BGM 小約翰 施特勞斯 藍色多瑙河 魷魚遊戲 Op.314 > 完成
轉換成功！輸出文件位於: C:/Users/Lu/Documents/youtube/yt_audio_The Beautiful Blue Danube .mp3


In [1]:
# V0.84 --> for test


from pytube import YouTube
import re
import subprocess

def get_audio_only(url):
    yt = YouTube(url)   
    print('全名 : ',yt.title)
    list=['/',':','"','<','>','\?','\*']
    yt_titles = yt.title
    for i in list:
        yt_title = re.sub(f'{i}','_',yt_titles)
        yt_titles = yt_title
    print(yt_titles)

    yt_title = re.search(r'[A-Za-z0-9_ ]+',yt_titles).group()
    print('簡稱 : ',yt_title)

    yt_audio = yt.streams.filter(abr='128kbps').first()
    yt_audio.download(output_path='C:/Users/Lu/Documents/youtube',filename=f'yt_audio_{yt_title}.mp4', skip_existing=True, max_retries=0)
    print(f'downdload < {yt.title} > 完成')
    return yt_title

def convert_mp4_to_mp3(input_file, output_file):
    command = [
        'ffmpeg',
        '-i', input_file,
        '-vn',
        '-acodec', 'libmp3lame',
        '-q:a', '3',
        output_file
    ]
    
    try:
        result = subprocess.run(command, check=True, stdout=subprocess.PIPE, stderr=subprocess.PIPE, text=True)
        print("轉換成功！輸出文件位於:", output_file)
    except subprocess.CalledProcessError as e:
        print("ffmpeg轉換過程中發生錯誤:", e)
        if e.stderr:
            print("ffmpeg命令輸出:", e.stderr)
        else:
            print("無法獲取 ffmpeg 錯誤輸出。")

url = 'https://www.youtube.com/watch?v=DkurWU1tDI0'  # 曲目:2024宮崎駿鋼琴音樂合集 - 最佳放鬆音樂《100%無廣告》スタジオジブリGhibli Relaxing Music【スターバックス】工作-讀書專用音樂

yt_title = get_audio_only(url)  #返回結果方式

convert_mp4_to_mp3(f'C:/Users/Lu/Documents/youtube/yt_audio_{yt_title}.mp4', f'C:/Users/Lu/Documents/youtube/yt_audio_{yt_title}.mp3')

# media 長度 01:34:55 mp4容量大小90,009KB mp3容量大小90,079KB 處理耗時2m 19.0s-->v0.84

全名 :  2時間のジブリ夏🍀🍀仕事勉強リラックスのためのジブリピアノBGM🍀少なくとも1 回 は 聞くべ き🍀Piano Studio Ghibli Collection
2時間のジブリ夏🍀🍀仕事勉強リラックスのためのジブリピアノBGM🍀少なくとも1 回 は 聞くべ き🍀Piano Studio Ghibli Collection
簡稱 :  2
downdload < 2時間のジブリ夏🍀🍀仕事勉強リラックスのためのジブリピアノBGM🍀少なくとも1 回 は 聞くべ き🍀Piano Studio Ghibli Collection > 完成
轉換成功！輸出文件位於: C:/Users/Lu/Documents/youtube/yt_audio_2.mp3


In [2]:
# V0.85
# 設定包含中英文字 re.search(r'[\w\u4e00-\u9fff\s]+',yt_titles)
#UnicodeDecodeError: 'cp950' codec can't decode byte 0xe6 in position 2196: illegal multibyte sequence

from pytube import YouTube
import re
import subprocess

def get_audio_only(url):
    yt = YouTube(url)   
    print('全名 : ',yt.title)
    list=['/',':','"','<','>','\?','\*']
    yt_titles = yt.title
    for i in list:
        yt_title = re.sub(f'{i}','_',yt_titles)
        yt_titles = yt_title
    print(yt_titles)

    #yt_title = re.search(r'[A-Za-z0-9_ ]+',yt_titles).group()
    yt_title = re.search(r'[\w\u4e00-\u9fff\s]+',yt_titles).group() #設定包含中英文字
    print('簡稱 : ',yt_title)

    yt_audio = yt.streams.filter(abr='128kbps').first()
    yt_audio.download(output_path='C:/Users/Lu/Documents/youtube',filename=f'yt_audio_{yt_title}.mp4', skip_existing=True, max_retries=0)
    print(f'downdload < {yt.title} > 完成')
    return yt_title

def convert_mp4_to_mp3(input_file, output_file):
    command = [
        'ffmpeg',
        '-i', input_file,
        '-vn',
        '-acodec', 'libmp3lame',
        '-q:a', '3',
        output_file
    ]
    
    try:
        result = subprocess.run(command, check=True, stdout=subprocess.PIPE, stderr=subprocess.PIPE, text=True)     # 無設定encoding='utf-8'
        print("轉換成功！輸出文件位於:", output_file)
    except subprocess.CalledProcessError as e:
        print("ffmpeg轉換過程中發生錯誤:", e)
        if e.stderr:
            print("ffmpeg命令輸出:", e.stderr)
        else:
            print("無法獲取 ffmpeg 錯誤輸出。")

url = 'https://www.youtube.com/watch?v=DkurWU1tDI0'  # 曲目:2024宮崎駿鋼琴音樂合集 - 最佳放鬆音樂《100%無廣告》スタジオジブリGhibli Relaxing Music【スターバックス】工作-讀書專用音樂

yt_title = get_audio_only(url)  #返回結果方式

convert_mp4_to_mp3(f'C:/Users/Lu/Documents/youtube/yt_audio_{yt_title}.mp4', f'C:/Users/Lu/Documents/youtube/yt_audio_{yt_title}.mp3')

# media 長度 01:34:55 mp4容量大小90,009KB mp3容量大小90,079KB 處理耗時2m 19.0s-->v0.85

全名 :  2時間のジブリ夏🍀🍀仕事勉強リラックスのためのジブリピアノBGM🍀少なくとも1 回 は 聞くべ き🍀Piano Studio Ghibli Collection
2時間のジブリ夏🍀🍀仕事勉強リラックスのためのジブリピアノBGM🍀少なくとも1 回 は 聞くべ き🍀Piano Studio Ghibli Collection
簡稱 :  2時間のジブリ夏
downdload < 2時間のジブリ夏🍀🍀仕事勉強リラックスのためのジブリピアノBGM🍀少なくとも1 回 は 聞くべ き🍀Piano Studio Ghibli Collection > 完成


Exception in thread Thread-9 (_readerthread):
Traceback (most recent call last):
  File "C:\python3.10.8\lib\threading.py", line 1016, in _bootstrap_inner
    self.run()
  File "c:\Users\Lu\Documents\GitHub\__2024_01_08_python_openAPI__\__youtube_download__\yt.venv\lib\site-packages\ipykernel\ipkernel.py", line 761, in run_closure
    _threading_Thread_run(self)
  File "C:\python3.10.8\lib\threading.py", line 953, in run
    self._target(*self._args, **self._kwargs)
  File "C:\python3.10.8\lib\subprocess.py", line 1499, in _readerthread
    buffer.append(fh.read())
UnicodeDecodeError: 'cp950' codec can't decode byte 0xe6 in position 2196: illegal multibyte sequence


轉換成功！輸出文件位於: C:/Users/Lu/Documents/youtube/yt_audio_2時間のジブリ夏.mp3


In [2]:
# V0.85 --> 2024.06.24 ok
# 設定包含中英文字 re.search(r'[\w\u4e00-\u9fff\s]+',yt_titles)
# subprocess.run設定encoding='utf-8'

from pytube import YouTube
import re
import subprocess

def get_audio_only(url):
    yt = YouTube(url)   
    print('全名 : ',yt.title)
    list=['/',':','"','<','>','\?','\*']
    yt_titles = yt.title
    for i in list:
        yt_title = re.sub(f'{i}','_',yt_titles)
        yt_titles = yt_title
    print(yt_titles)

    #yt_title = re.search(r'[A-Za-z0-9_ ]+',yt_titles).group()
    yt_title = re.search(r'[\w\u4e00-\u9fff\s]+',yt_titles).group() #設定包含中英文字
    print('簡稱 : ',yt_title)

    yt_audio = yt.streams.filter(abr='128kbps').first()
    yt_audio.download(output_path='C:/Users/Lu/Documents/youtube',filename=f'yt_audio_{yt_title}.mp4', skip_existing=True, max_retries=0)  
    print(f'downdload < {yt.title} > 完成')
    return yt_title

def convert_mp4_to_mp3(input_file, output_file):
    command = [
        'ffmpeg',
        '-i', input_file,
        '-vn',
        '-acodec', 'libmp3lame',
        '-q:a', '3',
        output_file
    ]
    
    try:
        result = subprocess.run(command, check=True, stdout=subprocess.PIPE, stderr=subprocess.PIPE, text=True,encoding= 'utf-8')  # 設定encoding='utf-8'
        print("轉換成功！輸出文件位於:", output_file)
    except subprocess.CalledProcessError as e:
        print("ffmpeg轉換過程中發生錯誤:", e)
        if e.stderr:
            print("ffmpeg命令輸出:", e.stderr)
        else:
            print("無法獲取 ffmpeg 錯誤輸出。")

url = 'https://www.youtube.com/watch?v=DkurWU1tDI0'  # 曲目:2024宮崎駿鋼琴音樂合集 - 最佳放鬆音樂《100%無廣告》スタジオジブリGhibli Relaxing Music【スターバックス】工作-讀書專用音樂

yt_title = get_audio_only(url)  #返回結果方式

convert_mp4_to_mp3(f'C:/Users/Lu/Documents/youtube/yt_audio_{yt_title}.mp4', f'C:/Users/Lu/Documents/youtube/yt_audio_{yt_title}.mp3')

# media 長度 01:34:55 mp4容量大小90,009KB mp3容量大小90,079KB 處理耗時2m 14.1s-->v0.85

全名 :  2時間のジブリ夏🍀🍀仕事勉強リラックスのためのジブリピアノBGM🍀少なくとも1 回 は 聞くべ き🍀Piano Studio Ghibli Collection
2時間のジブリ夏🍀🍀仕事勉強リラックスのためのジブリピアノBGM🍀少なくとも1 回 は 聞くべ き🍀Piano Studio Ghibli Collection
簡稱 :  2時間のジブリ夏
downdload < 2時間のジブリ夏🍀🍀仕事勉強リラックスのためのジブリピアノBGM🍀少なくとも1 回 は 聞くべ き🍀Piano Studio Ghibli Collection > 完成
轉換成功！輸出文件位於: C:/Users/Lu/Documents/youtube/yt_audio_2時間のジブリ夏.mp3


In [9]:
# V0.9 --> 2024.06.24 final

from pytube import YouTube
import re
import subprocess

def streams_analysis(url):
    yt = YouTube(url) 
    # 影片格式分析
    print(f'標題: {yt.title}')              # 影片標題
    print(f'長度(s): {yt.length}')          # 影片長度 ( 秒 )
    print(f'作者: {yt.author}')             # 影片作者
    print(f'網址: {yt.channel_url}')        # 影片作者頻道網址
    print(f'縮圖網址: {yt.thumbnail_url}')  # 影片縮圖網址
    print(f'觀看數: {yt.views}')            # 影片觀看數

    # 分析
    print(f'支援畫質:{yt.streams}')  # 影片支援哪些畫質
    # print(yt.streams.all())  # all()-->useless
    print('影片支援畫質數: ',len(yt.streams))

    for st in yt.streams:
        print(st)


def get_audio_only(url):
    yt = YouTube(url)   
    print('全名 : ',yt.title)
    list=['/',':','"','<','>','\?','\*']
    yt_titles = yt.title
    for i in list:
        yt_title = re.sub(f'{i}','_',yt_titles)
        yt_titles = yt_title
    print(yt_titles)

    #yt_title = re.search(r'[A-Za-z0-9_ ]+',yt_titles).group()
    yt_title = re.search(r'[\w\u4e00-\u9fff\s]+',yt_titles).group() #設定包含中英文字
    print('簡稱 : ',yt_title)

    yt_audio = yt.streams.filter(abr='128kbps').first()
    yt_audio.download(output_path='C:/Users/Lu/Documents/youtube',filename=f'yt_audio_{yt_title}.mp4', skip_existing=True, max_retries=0)  
    print(f'downdload < {yt.title} > 完成')
    return yt_title

def convert_mp4_to_mp3(input_file, output_file):
    command = [
        'ffmpeg',
        '-i', input_file,
        '-vn',
        '-acodec', 'libmp3lame',
        '-q:a', '3',
        output_file
    ]
    
    try:
        result = subprocess.run(command, check=True, stdout=subprocess.PIPE, stderr=subprocess.PIPE, text=True,encoding= 'utf-8')  # 設定encoding='utf-8'
        print("轉換成功！輸出文件位於:", output_file)
    except subprocess.CalledProcessError as e:
        print("ffmpeg轉換過程中發生錯誤:", e)
        if e.stderr:
            print("ffmpeg命令輸出:", e.stderr)
        else:
            print("無法獲取 ffmpeg 錯誤輸出。")

 
# url = 'https://www.youtube.com/watch?v=OmLqW4DQCDw'  # 曲目:Walt Disney Songs Collection with Lyrics 2024 🛕 The Most Romantic Disney Songs - Disney Soundtracks

work_item = int(input('輸入 ( 1 :分析   2:下載   其他:跳出執行 )'))
if work_item == 1:
   url = str(input("輸入影音網址 (Enter the url of media) : ")) 
   streams_analysis(url) 
elif work_item == 2:
    url = str(input("輸入影音網址(Enter the url of media) : "))  
    yt_title = get_audio_only(url)  #返回結果方式
    convert_mp4_to_mp3(f'C:/Users/Lu/Documents/youtube/yt_audio_{yt_title}.mp4', f'C:/Users/Lu/Documents/youtube/yt_audio_{yt_title}.mp3')
else:
     print('輸入 ( 1 :分析   2:下載   其他:跳出執行 )')


# media 長度 04:59:43 mp4容量大小284,226KB mp3容量大小341,219KB 處理耗時6m 27.1s-->v0.9

全名 :  Walt Disney Songs Collection with Lyrics 2024 🛕 The Most Romantic Disney Songs - Disney Soundtracks
Walt Disney Songs Collection with Lyrics 2024 🛕 The Most Romantic Disney Songs - Disney Soundtracks
簡稱 :  Walt Disney Songs Collection with Lyrics 2024 
downdload < Walt Disney Songs Collection with Lyrics 2024 🛕 The Most Romantic Disney Songs - Disney Soundtracks > 完成
轉換成功！輸出文件位於: C:/Users/Lu/Documents/youtube/yt_audio_Walt Disney Songs Collection with Lyrics 2024 .mp3
